zebra/zebra_mpls.c

```diff
@@ -2631,141 +2812,391 @@ int mpls_ftn_update(int add, struct zebra_vrf *zvrf, enum lsp_types_t type,
+		struct nexthop *nh = nhlfe->nexthop;
 
-	/* Lookup existing route */
-	rn = route_node_get(table, prefix);
-	RNODE_FOREACH_RE (rn, re) {
-		if (CHECK_FLAG(re->status, ROUTE_ENTRY_REMOVED))
-			continue;
-		if (re->type == route_type && re->instance == route_instance)
-			break;
-	}
-	if (re == NULL)
-		return -1;
+		assert(nh);
+		assert(nh->nh_label);
 
-	nexthop_group_copy(&new_grp, &(re->nhe->nhg));
+		/* Clear deleted flag (in case it was set) */
+		UNSET_FLAG(nhlfe->flags, NHLFE_FLAG_DELETED);
+		if (nh->nh_label->num_labels == num_out_labels
+		    && !memcmp(nh->nh_label->label, out_labels,
+			       sizeof(mpls_label_t) * num_out_labels))
+			/* No change */
+			return nhlfe;
 
-	for (nexthop = new_grp.nexthop; nexthop; nexthop = nexthop->next)
-		nexthop_del_labels(nexthop);
+		if (IS_ZEBRA_DEBUG_MPLS) {
+			char buf2[MPLS_LABEL_STRLEN];
+			char buf3[MPLS_LABEL_STRLEN];
 
-	SET_FLAG(re->status, ROUTE_ENTRY_CHANGED);
-	SET_FLAG(re->status, ROUTE_ENTRY_LABELS_CHANGED);
+			nhlfe2str(nhlfe, buf, sizeof(buf));
+			mpls_label2str(num_out_labels, out_labels, buf2,
+				       sizeof(buf2), 0);
+			mpls_label2str(nh->nh_label->num_labels,
+				       nh->nh_label->label, buf3, sizeof(buf3),
+				       0);
+
+			zlog_debug("LSP in-label %u type %d nexthop %s out-label(s) changed to %s (old %s)",
+				   lsp->ile.in_label, type, buf, buf2, buf3);
+		}
+
+		/* Update out label(s), trigger processing. */
+		if (nh->nh_label->num_labels == num_out_labels)
+			memcpy(nh->nh_label->label, out_labels,
+			       sizeof(mpls_label_t) * num_out_labels);
+		else {
+			nexthop_del_labels(nh);
+			nexthop_add_labels(nh, type, num_out_labels,
+					   out_labels);
+		}
+	} else {
+		/* Add LSP entry to this nexthop */
+		nhlfe = nhlfe_add(lsp, type, gtype, gate, ifindex,
+				  num_out_labels, out_labels);
+		if (!nhlfe)
+			return NULL;
 
-	mpls_zebra_nhg_update(re, afi, &new_grp);
+		if (IS_ZEBRA_DEBUG_MPLS) {
+			char buf2[MPLS_LABEL_STRLEN];
 
-	nexthops_free(new_grp.nexthop);
+			nhlfe2str(nhlfe, buf, sizeof(buf));
+			mpls_label2str(num_out_labels, out_labels, buf2,
+				       sizeof(buf2), 0);
 
-	rib_queue_add(rn);
+			zlog_debug("Add LSP in-label %u type %d nexthop %s out-label(s) %s",
+				   lsp->ile.in_label, type, buf, buf2);
+		}
 
-	return 0;
+		lsp->addr_family = NHLFE_FAMILY(nhlfe);
+	}
+
+	/* Mark NHLFE, queue LSP for processing. */
+	SET_FLAG(nhlfe->flags, NHLFE_FLAG_CHANGED);
+
+	return nhlfe;
 }
 
 /*
  * Install/update a NHLFE for an LSP in the forwarding table. This may be
  * a new LSP entry or a new NHLFE for an existing in-label or an update of
  * the out-label for an existing NHLFE (update case).
  */
-int mpls_lsp_install(struct zebra_vrf *zvrf, enum lsp_types_t type,
-		     mpls_label_t in_label, uint8_t num_out_labels,
-		     mpls_label_t out_labels[], enum nexthop_types_t gtype,
-		     const union g_addr *gate, ifindex_t ifindex)
+static zebra_nhlfe_t *
+lsp_add_backup_nhlfe(zebra_lsp_t *lsp, enum lsp_types_t type,
```

Comment:
nit: this function and `lsp_add_nhlfe()` are almost identical except for a few small differences. Wouldn't it be possible to add a new "backup" parameter to `lsp_add_nhlfe()` to avoid the code duplication?

The same comment applies to:
- `mpls_lsp_uninstall()` and `lsp_backup_uninstall()`;
- `nhlfe_add()` and `nhlfe_backup_add()`.

Reply:
yes, you're right - I struggled with that question a bit, and yes, I think I can eliminate some duplication. would you like that done with this PR now, or could I include that in the next 'phase' PR ?

Reply:
Yeah sure. I should have said that was a non-blocking comment (both of them actually).

Marking as resolved...